File: config-manager.js
/**
 * Configuration Manager for Aditum Gate System (Node.js)
 * Fetches configuration from external API and manages local caching
 */

const fs = require('fs');
const path = require('path');
const http = require('http');
const https = require('https');

// Get the directory where config-manager.js is located
const CONFIG_MANAGER_DIR = __dirname;

const DEVICE_ID_FILE = path.join(CONFIG_MANAGER_DIR, 'device-id.txt');
const CONFIG_CACHE_FILE = path.join(CONFIG_MANAGER_DIR, 'config-cache.json');
const CONFIG_DEFAULT_FILE = path.join(CONFIG_MANAGER_DIR, 'config-default.json');

let _configCache = null;
let _pollingInterval = null;
let _lastConfigHash = null;

/**
 * Read device ID from device-id.txt file
 */
function getDeviceId() {
    try {
        if (fs.existsSync(DEVICE_ID_FILE)) {
            const deviceId = fs.readFileSync(DEVICE_ID_FILE, 'utf8').trim();
            if (deviceId) {
                return deviceId;
            }
        }
    } catch (error) {
        console.error(`Error reading device ID: ${error.message}`);
    }
    
    return 'DEVICE-001';
}

/**
 * Load default configuration from config-default.json
 */
function loadDefaultConfig() {
    try {
        if (fs.existsSync(CONFIG_DEFAULT_FILE)) {
            const content = fs.readFileSync(CONFIG_DEFAULT_FILE, 'utf8');
            return JSON.parse(content);
        }
    } catch (error) {
        console.error(`Error loading default config: ${error.message}`);
    }
    
    // Return minimal default config if file doesn't exist
    return {
        device: { deviceId: 'DEVICE-001', deviceName: 'Default Device', scannerType: 'qr', scannerScript: 'scannerQr.py' },
        hardware: { hasScreen: true, hasTwoCameras: true, isScreen: true, deviceName: 'Newtologic  4010E' },
        door: { doorType: 'entry', doorId: '0', placeName: 'Name' },
        display: { clientLogoUrl: 'https://res.cloudinary.com/aditum/image/upload/v1501920877/fzncrputkdgm8iasuc3t.jpg', showCameraFeed: true },
        api: { baseUrl: 'https://app.aditumcr.com/api' },
        polling: { intervalSeconds: 30, enabled: true }
    };
}

/**
 * Load configuration from local cache file
 */
function loadCachedConfig() {
    try {
        if (fs.existsSync(CONFIG_CACHE_FILE)) {
            const content = fs.readFileSync(CONFIG_CACHE_FILE, 'utf8');
            return JSON.parse(content);
        }
    } catch (error) {
        console.error(`Error loading cached config: ${error.message}`);
    }
    
    return null;
}

/**
 * Save configuration to local cache file
 */
function saveConfigCache(config) {
    try {
        fs.writeFileSync(CONFIG_CACHE_FILE, JSON.stringify(config, null, 2), 'utf8');
    } catch (error) {
        console.error(`Error saving config cache: ${error.message}`);
    }
}

/**
 * Fetch configuration from external API
 */
function fetchConfigFromApi(deviceId, baseUrl = null) {
    return new Promise((resolve, reject) => {
        if (!baseUrl) {
            const defaultConfig = loadDefaultConfig();
            baseUrl = defaultConfig.api?.baseUrl || 'https://app.aditumcr.com/api';
        }
        
        // Derive config API URL from baseUrl (remove /api if present, then add /api/devices/{id}/config)
        const urlObj = new URL(baseUrl);
        const baseDomain = `${urlObj.protocol}//${urlObj.host}`;
        const url = `${baseDomain}/api/devices/${deviceId}/config`;
        const urlObj = new URL(url);
        const isHttps = urlObj.protocol === 'https:';
        const httpModule = isHttps ? https : http;
        
        const options = {
            hostname: urlObj.hostname,
            port: urlObj.port || (isHttps ? 443 : 80),
            path: urlObj.pathname + urlObj.search,
            method: 'GET',
            timeout: 10000
        };
        
        const req = httpModule.request(options, (res) => {
            let data = '';
            
            res.on('data', (chunk) => {
                data += chunk;
            });
            
            res.on('end', () => {
                if (res.statusCode === 200) {
                    try {
                        resolve(JSON.parse(data));
                    } catch (error) {
                        reject(new Error(`Failed to parse JSON: ${error.message}`));
                    }
                } else {
                    reject(new Error(`API returned status ${res.statusCode}`));
                }
            });
        });
        
        req.on('error', (error) => {
            reject(error);
        });
        
        req.on('timeout', () => {
            req.destroy();
            reject(new Error('Request timeout'));
        });
        
        req.end();
    });
}

/**
 * Compare two configs and return list of changed field paths
 */
function compareConfigs(oldConfig, newConfig) {
    if (!oldConfig) {
        return null;
    }
    
    const changedFields = [];
    
    function compareObjects(old, newObj, currentPath = '') {
        if (typeof old !== 'object' || typeof newObj !== 'object' || old === null || newObj === null) {
            if (old !== newObj) {
                changedFields.push(currentPath || 'root');
            }
            return;
        }
        
        const allKeys = new Set([...Object.keys(old), ...Object.keys(newObj)]);
        for (const key of allKeys) {
            const path = currentPath ? `${currentPath}.${key}` : key;
            if (!(key in old)) {
                changedFields.push(path);
            } else if (!(key in newObj)) {
                changedFields.push(path);
            } else {
                if (typeof old[key] === 'object' && typeof newObj[key] === 'object' && old[key] !== null && newObj[key] !== null) {
                    compareObjects(old[key], newObj[key], path);
                } else if (old[key] !== newObj[key]) {
                    changedFields.push(path);
                }
            }
        }
    }
    
    compareObjects(oldConfig, newConfig);
    return changedFields.length > 0 ? changedFields : null;
}

/**
 * Update configuration from external API
 */
async function updateConfig() {
    const deviceId = getDeviceId();
    const defaultConfig = loadDefaultConfig();
    const baseUrl = defaultConfig.api?.baseUrl || 'https://app.aditumcr.com/api';
    
    try {
        const newConfig = await fetchConfigFromApi(deviceId, baseUrl);
        
        if (!newConfig) {
            // API unavailable, try to use cached config
            const cachedConfig = loadCachedConfig();
            if (cachedConfig) {
                console.log('Using cached config (API unavailable)');
                _configCache = cachedConfig;
                return;
            } else {
                // No cache, use default
                console.log('Using default config (API unavailable, no cache)');
                _configCache = defaultConfig;
                saveConfigCache(defaultConfig);
                return;
            }
        }
        
        // Check if config changed
        const oldConfig = _configCache;
        const configHash = JSON.stringify(newConfig);
        
        if (!_lastConfigHash || configHash !== _lastConfigHash) {
            const changedFields = compareConfigs(oldConfig, newConfig);
            
            _configCache = newConfig;
            _lastConfigHash = configHash;
            
            saveConfigCache(newConfig);
            
            if (changedFields) {
                console.log(`Configuration updated. Changed fields: ${changedFields.join(', ')}`);
            } else {
                console.log('Configuration reloaded (first load)');
            }
        } else {
            console.log('Configuration unchanged');
        }
    } catch (error) {
        console.error(`Error updating config: ${error.message}`);
        
        // Fallback to cached or default
        const cachedConfig = loadCachedConfig();
        if (cachedConfig) {
            _configCache = cachedConfig;
        } else {
            _configCache = defaultConfig;
            saveConfigCache(defaultConfig);
        }
    }
}

/**
 * Start background polling for config updates
 */
function startPolling() {
    if (_pollingInterval) {
        return; // Already polling
    }
    
    const defaultConfig = loadDefaultConfig();
    const pollingConfig = defaultConfig.polling || {};
    const interval = (pollingConfig.intervalSeconds || 30) * 1000;
    
    if (pollingConfig.enabled !== false) {
        _pollingInterval = setInterval(() => {
            updateConfig().catch(error => {
                console.error(`Error in polling: ${error.message}`);
            });
        }, interval);
        
        console.log('Configuration polling started');
    }
}

/**
 * Stop background polling
 */
function stopPolling() {
    if (_pollingInterval) {
        clearInterval(_pollingInterval);
        _pollingInterval = null;
        console.log('Configuration polling stopped');
    }
}

/**
 * Get current configuration (returns a copy)
 */
function getConfig() {
    if (!_configCache) {
        // First load - try cache, then default
        const cached = loadCachedConfig();
        if (cached) {
            _configCache = cached;
        } else {
            _configCache = loadDefaultConfig();
            saveConfigCache(_configCache);
        }
    }
    
    // Return a deep copy to prevent external modifications
    return JSON.parse(JSON.stringify(_configCache));
}

/**
 * Initialize configuration system
 */
async function initConfig() {
    await updateConfig();
    startPolling();
}

// Auto-initialize when module is loaded
if (require.main !== module) {
    initConfig().catch(error => {
        console.error(`Error initializing config: ${error.message}`);
    });
}

module.exports = {
    getConfig,
    getDeviceId,
    initConfig,
    updateConfig,
    startPolling,
    stopPolling
};

File: config-manager.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Configuration Manager for Aditum Gate System
Fetches configuration from external API and manages local caching
"""

import json
import os
import time
import threading
import requests
from urllib.parse import urlparse

# File paths
DEVICE_ID_FILE = 'device-id.txt'
CONFIG_CACHE_FILE = 'config-cache.json'
CONFIG_DEFAULT_FILE = 'config-default.json'

# Global config cache
_config_cache = None
_config_lock = threading.Lock()
_last_config_hash = None
_polling_thread = None
_polling_active = False


def get_device_id():
    """Read device ID from device-id.txt file"""
    try:
        if os.path.exists(DEVICE_ID_FILE):
            with open(DEVICE_ID_FILE, 'r') as f:
                device_id = f.read().strip()
                if device_id:
                    return device_id
    except Exception as e:
        print(f"Error reading device ID: {e}")
    
    # Return default if file doesn't exist or is empty
    return "DEVICE-001"


def load_default_config():
    """Load default configuration from config-default.json"""
    try:
        if os.path.exists(CONFIG_DEFAULT_FILE):
            with open(CONFIG_DEFAULT_FILE, 'r') as f:
                return json.load(f)
    except Exception as e:
        print(f"Error loading default config: {e}")
    
    # Return minimal default config if file doesn't exist
    return {
        "device": {"deviceId": "DEVICE-001", "deviceName": "Default Device", "scannerType": "qr", "scannerScript": "scannerQr.py"},
        "hardware": {"hasScreen": True, "hasTwoCameras": True, "isScreen": True, "deviceName": "Newtologic  4010E"},
        "door": {"doorType": "entry", "doorId": "0", "placeName": "Name"},
        "display": {"clientLogoUrl": "https://res.cloudinary.com/aditum/image/upload/v1501920877/fzncrputkdgm8iasuc3t.jpg", "showCameraFeed": True},
        "api": {"baseUrl": "https://app.aditumcr.com/api"},
        "polling": {"intervalSeconds": 30, "enabled": True}
    }


def load_cached_config():
    """Load configuration from local cache file"""
    try:
        if os.path.exists(CONFIG_CACHE_FILE):
            with open(CONFIG_CACHE_FILE, 'r') as f:
                return json.load(f)
    except Exception as e:
        print(f"Error loading cached config: {e}")
    
    return None


def save_config_cache(config):
    """Save configuration to local cache file"""
    try:
        with open(CONFIG_CACHE_FILE, 'w') as f:
            json.dump(config, f, indent=2)
    except Exception as e:
        print(f"Error saving config cache: {e}")


def fetch_config_from_api(device_id, base_url=None):
    """Fetch configuration from external API"""
    if not base_url:
        # Try to get from default config first
        default_config = load_default_config()
        base_url = default_config.get('api', {}).get('baseUrl', 'https://app.aditumcr.com/api')
    
    # Derive config API URL from baseUrl (remove /api if present, then add /api/devices/{id}/config)
    parsed = urlparse(base_url)
    base_domain = f"{parsed.scheme}://{parsed.netloc}"
    url = f"{base_domain}/api/devices/{device_id}/config"
    
    try:
        response = requests.get(url, timeout=10)
        if response.status_code == 200:
            return response.json()
        else:
            print(f"API returned status {response.status_code} for config fetch")
            return None
    except requests.exceptions.RequestException as e:
        print(f"Error fetching config from API: {e}")
        return None


def compare_configs(old_config, new_config):
    """Compare two configs and return list of changed field paths"""
    if old_config is None:
        return None
    
    changed_fields = []
    
    def compare_dicts(old, new, path=""):
        if not isinstance(old, dict) or not isinstance(new, dict):
            if old != new:
                changed_fields.append(path if path else "root")
            return
        
        all_keys = set(old.keys()) | set(new.keys())
        for key in all_keys:
            current_path = f"{path}.{key}" if path else key
            if key not in old:
                changed_fields.append(current_path)
            elif key not in new:
                changed_fields.append(current_path)
            else:
                if isinstance(old[key], dict) and isinstance(new[key], dict):
                    compare_dicts(old[key], new[key], current_path)
                elif old[key] != new[key]:
                    changed_fields.append(current_path)
    
    compare_dicts(old_config, new_config)
    return changed_fields if changed_fields else None


def update_config():
    """Fetch and update configuration from external API"""
    global _config_cache, _last_config_hash
    
    device_id = get_device_id()
    default_config = load_default_config()
    base_url = default_config.get('api', {}).get('baseUrl', 'https://app.aditumcr.com/api')
    
    # Try to fetch from API
    new_config = fetch_config_from_api(device_id, base_url)
    
    if new_config is None:
        # API unavailable, try to use cached config
        cached_config = load_cached_config()
        if cached_config:
            print("Using cached config (API unavailable)")
            with _config_lock:
                _config_cache = cached_config
            return
        else:
            # No cache, use default
            print("Using default config (API unavailable, no cache)")
            with _config_lock:
                _config_cache = default_config
            save_config_cache(default_config)
            return
    
    # Check if config changed
    old_config = _config_cache
    config_hash = hash(json.dumps(new_config, sort_keys=True))
    
    if _last_config_hash is None or config_hash != _last_config_hash:
        changed_fields = compare_configs(old_config, new_config)
        
        with _config_lock:
            _config_cache = new_config
            _last_config_hash = config_hash
        
        save_config_cache(new_config)
        
        if changed_fields:
            print(f"Configuration updated. Changed fields: {', '.join(changed_fields)}")
        else:
            print("Configuration reloaded (first load)")
    else:
        print("Configuration unchanged")


def polling_worker():
    """Background worker that polls for config updates"""
    global _polling_active
    
    default_config = load_default_config()
    polling_config = default_config.get('polling', {})
    interval = polling_config.get('intervalSeconds', 30)
    
    while _polling_active:
        try:
            if polling_config.get('enabled', True):
                update_config()
        except Exception as e:
            print(f"Error in polling worker: {e}")
        
        time.sleep(interval)


def start_polling():
    """Start background polling thread"""
    global _polling_thread, _polling_active
    
    if _polling_thread is None or not _polling_thread.is_alive():
        _polling_active = True
        _polling_thread = threading.Thread(target=polling_worker, daemon=True)
        _polling_thread.start()
        print("Configuration polling started")


def stop_polling():
    """Stop background polling thread"""
    global _polling_active
    
    _polling_active = False
    if _polling_thread:
        _polling_thread.join(timeout=5)
    print("Configuration polling stopped")


def get_config():
    """Get current configuration (thread-safe)"""
    global _config_cache
    
    with _config_lock:
        if _config_cache is None:
            # First load - try cache, then default
            cached = load_cached_config()
            if cached:
                _config_cache = cached
            else:
                _config_cache = load_default_config()
                save_config_cache(_config_cache)
        
        # Return a copy to prevent external modifications
        return json.loads(json.dumps(_config_cache))


def init_config():
    """Initialize configuration system - fetch from API and start polling"""
    update_config()
    start_polling()


# Initialize on import
if __name__ != "__main__":
    # Auto-initialize when imported
    init_config()